### PR TITLE
Remote inference from service endpoint

### PR DESCRIPTION
**Pull request by kamilwu**
This PR adds a `_RemotePredictDoFn` which performs predictions from a model hosted on Google Cloud AI Platform.

Tested on Dataflow runner in both batch and streaming modes.

### CODE REVIEW

**Comment by kamilwu** [comment]
R: @brills @rose-rong-liu @zhitaoli I'd be happy to get a review from you.
cc: @Ardagan

File: tfx_bsl/beam/run_inference.py [hunk 1]
```diff
@@ -71,27 +73,26 @@
...
+  CLASSIFICATION = 'CLASSIFICATION'
+  REGRESSION = 'REGRESSION'
+  PREDICTION = 'PREDICTION'
+  MULTIHEAD = 'MULTIHEAD'
+
+
 @beam.ptransform_fn
 @beam.typehints.with_input_types(Union[tf.train.Example,
                                        tf.train.SequenceExample])
-@beam.typehints.with_output_types(prediction_log_pb2.PredictionLog)
```
Comment (zhitaoli) [inline tfx_bsl/beam/run_inference.py]: Why is this type hint removed? Are we not emitting this type as an output anymore

Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: Not always. If we use remote inference, the transform will emit a response body as an output. The response body, in turn, contains a collection of predictions which is a JSON list[1]. That's why I removed the hint.

[1] https://cloud.google.com/ai-platform/prediction/docs/reference/rest/v1/projects/predict#response-body-details

**Comment by kamilwu** [comment]
@googlebot I signed it!

File: tfx_bsl/beam/run_inference.py [hunk 2]
```diff
@@ -71,27 +73,26 @@
                                classification_pb2.Classifications]]
 
 
+class OperationType(object):
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Inherit from Enum?

File: tfx_bsl/beam/run_inference.py [hunk 3]
```diff
@@ -71,27 +73,26 @@
...
     inference_endpoint: model_spec_pb2.InferenceEndpoint
 ) -> beam.pvalue.PCollection:
-  """Run batch offline inference with a model.
-
-  Models need to have the required serving signature as mentioned in
-  [Tensorflow Serving](https://www.tensorflow.org/tfx/serving/signature_defs)
-
-  This function will check model signatures first. Then it will load and run
-  model inference in batch.
+  """Run inference with a model.
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Could you add comments about how to use saved_model_spec and model_endpoint_spec?

File: tfx_bsl/beam/run_inference.py [hunk 4]
```diff
@@ -799,6 +900,13 @@ def _signature_pre_process_regress(
   return input_tensor_name, output_alias_tensor_names
 
 
+def _using_offline_inference(inference_endpoint:
+  model_spec_pb2.InferenceEndpoint) -> bool:
+  return isinstance(getattr(inference_endpoint,
+                            inference_endpoint.WhichOneof('type')),
+                    model_spec_pb2.SavedModelSpec)
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: You can use inference_endpoint.WhichOneof("type") == "saved_model_spec"

File: tfx_bsl/beam/run_inference.py [hunk 5]
```diff
@@ -221,17 +223,129 @@ def update(self, elements: List[Union[tf.train.Example,
...
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> list:
+    raise NotImplementedError
+
+
+@beam.typehints.with_input_types(List[Union[tf.train.Example,
+                                            tf.train.SequenceExample]])
+class _RemotePredictDoFn(_BaseDoFn):
+  """A DoFn that performs predictions from a cloud-hosted model.
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Is it limited to cloud-hosted tensorflow model? If so, please mention it in comment.

File: tfx_bsl/beam/run_inference.py [hunk 6]
```diff
@@ -221,17 +223,129 @@ def update(self, elements: List[Union[tf.train.Example,
...
+  @staticmethod
+  def _execute_request(request: http.HttpRequest) -> dict:
+    return request.execute()
+
+  def _make_request(self, model_name: str, body: dict) -> http.HttpRequest:
+    return self._api_client.projects().predict(name=model_name, body=body)
+
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> list:
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: What is the normal return value from this function?

File: tfx_bsl/beam/run_inference.py [hunk 7]
```diff
@@ -221,17 +223,129 @@ def update(self, elements: List[Union[tf.train.Example,
...
+    batch_start_time = self._clock.get_current_time_in_microseconds()
+    result = self.run_inference(elements)
+    self._metrics_collector.update(
+        elements,
+        self._clock.get_current_time_in_microseconds() - batch_start_time)
+    return result
+
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> list:
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Could this be a pytype type?
For saved_model_spec, this is Mapping[Text, np.ndarray]. If you need to extend it to a list, you can use Union[List,...]

File: tfx_bsl/beam/run_inference.py (hunk 2 above)
Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: I was concerned about the lack of `Enum` on Python 2.7, but we already have access to `enum34` (backported Enum), since this is a dependency of Beam. If you are fine with marking `enum34` as tfx_bsl`s dependency, I think I can use it.

File: tfx_bsl/beam/run_inference.py (hunk 5 above)
Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: Yes, it's limited to tensorflow model. I'll add a mention.

File: tfx_bsl/beam/run_inference.py (hunk 6 above)
Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: It returns a list of JSON objects which could be of any complexity: it depends on model. For instance, according to docs, if model has more than one output tensor, each element in the list is a dictionary containing name/value pairs.

I'll try to convert the response to PredictLog objects.

File: tfx_bsl/beam/run_inference.py (hunk 2 above)
Comment (rose-rong-liu) [reply tfx_bsl/beam/run_inference.py]: Ah, good point. enum34 is not a necessary dependency to add. Let's keep it this way.

**Comment by kamilwu** [comment]
@zhitaoli  @rose-rong-liu  Thanks for the feedback!

I've improved documentation and fixed one mistake in unit tests. I've also improved building request body: now if the name of an input ends in `_bytes`, it's a signal for us to treat data as a binary data.

Please take a look once again.

File: tfx_bsl/beam/run_inference.py [hunk 8]
```diff
@@ -221,17 +232,207 @@ def update(self, elements: List[Union[tf.train.Example,
...
+
+  def _make_request(self, model_name: str, body: dict) -> http.HttpRequest:
+    return self._api_client.projects().predict(name=model_name, body=body)
+
+  @classmethod
+  def _prepare_instances(cls,
+    elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Generator[Mapping[Text, Any], None, None]:
+    for example in elements:
+      # TODO: support tf.train.SequenceExample
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: If Sequence Example is not supported, maybe also reflect it on the pytype of line 320?

File: tfx_bsl/beam/run_inference.py [hunk 9]
```diff
@@ -221,17 +232,207 @@ def update(self, elements: List[Union[tf.train.Example,
...
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Sequence[Mapping[Text, Any]]:
+    body = {
+      'instances': list(self._prepare_instances(elements))
+    }
+    response = self._execute_request(self._make_request(self.full_model_name,
+                                                        body))
+    try:
+      error_msg = response['error']
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Is response a map? If so, you can use:
if 'error' in response

File: tfx_bsl/beam/run_inference.py [hunk 10]
```diff
@@ -221,17 +232,207 @@ def update(self, elements: List[Union[tf.train.Example,
...
+        values = cls._parse_feature_content(attr.value, attr_name,
+                                            cls._sending_as_binary(input_name))
+        values = cls._maybe_flatten(values)
+        instance[input_name] = values
+      yield instance
+
+  @staticmethod
+  def _sending_as_binary(input_name: Text) -> bool:
+    """Whether data should be sent as binary."""
+    return input_name.endswith('_bytes')
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: Could you add the description of "_bytes" suffix to the _RemotePredictDoFn comment? Thanks

Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: Done. Thanks!

File: tfx_bsl/beam/run_inference.py (hunk 9 above)
Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: Done. It looks a bit better after that the change.

**Comment by kamilwu** [comment]
@rose-rong-liu Thanks. I've pushed new fixes.

As for the returning `PredictLog` instances as a result, I have a question.

For now, my implementation doesn't fill `PredictRequest` part of the `PredictLog` message. As we are using cloud-hosted model, we don't have access to its signatures. That's why it's difficult to determine e.g. the input tensor name/alias. 

What should we do about this?

**Comment by rose-rong-liu** [comment]
> @rose-rong-liu Thanks. I've pushed new fixes.
> 
> As for the returning `PredictLog` instances as a result, I have a question.
> 
> For now, my implementation doesn't fill `PredictRequest` part of the `PredictLog` message. As we are using cloud-hosted model, we don't have access to its signatures. That's why it's difficult to determine e.g. the input tensor name/alias.
> 
> What should we do about this?

ICould you leave a comment on _RemotePredictDoFn mentioning this? Thanks

File: tfx_bsl/beam/run_inference.py [hunk 11]
```diff
@@ -221,17 +232,209 @@ def update(self, elements: List[Union[tf.train.Example,
...
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Sequence[Mapping[Text, Any]]:
+    body = {
+      'instances': list(self._prepare_instances(elements))
+    }
+    response = self._execute_request(self._make_request(self.full_model_name,
+                                                        body))
+    if 'error' in response:
+      raise ValueError(response['error'])
+    else:
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/run_inference.py]: "else" is not needed, just return is enough :)

**Comment by kamilwu** [comment]
@rose-rong-liu Thanks for the review.
@zhitaoli (and others as well) Is there anything else I need to take care of?

**Comment by brills** [comment]
There was no "cla: yes" label at the time the author signed the CLA (https://github.com/tensorflow/tfx-bsl/pull/2#issuecomment-588265299). I'm manually attached the label.

File: tfx_bsl/beam/run_inference.py (hunk 2 above)
Comment (zhitaoli) [reply tfx_bsl/beam/run_inference.py]: Let's have a TODO to convert this into enum once we stopped supporting python 2.7, which should be sometime Q2-Q3 this year.

Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: Done, a TODO has been added.

File: tfx_bsl/beam/run_inference.py (hunk 1 above)
Comment (Ardagan) [reply tfx_bsl/beam/run_inference.py]: Do I understand correctly that depending on usage, we get different output types here?
If yes, can we standardize output?

File: tfx_bsl/beam/run_inference.py [hunk 12]
```diff
@@ -221,17 +233,212 @@ def update(self, elements: List[Union[tf.train.Example,
...
+  def _post_process(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]],
+      outputs: Any
+  ) -> Iterable[Any]:
+    raise NotImplementedError
+
+
+@beam.typehints.with_input_types(List[Union[tf.train.Example,
+                                            tf.train.SequenceExample]])
+# Using output typehints causes an error on streaming mode on Dataflow runner.
```
Comment (Ardagan) [inline tfx_bsl/beam/run_inference.py]: Can you please elaborate on what causes error and provide a link if this is a bug?

Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: It hits the following exception:
```
  def to_type_hint(self):
    raise NotImplementedError('BEAM-2717')
```
in file: sdks/python/apache_beam/coders/coders.py
A link to the issue: https://issues.apache.org/jira/browse/BEAM-2717

Strangely enough, eveything works in Batch mode on both Dataflow and DirectRunner. 
I'll add issue ID to the comment.

File: tfx_bsl/beam/run_inference.py [hunk 13]
```diff
@@ -221,17 +233,213 @@ def update(self, elements: List[Union[tf.train.Example,
...
+  def _post_process(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]],
+      outputs: Any
+  ) -> Iterable[Any]:
+    raise NotImplementedError
+
+
+@beam.typehints.with_input_types(List[Union[tf.train.Example,
+                                            tf.train.SequenceExample]])
+# Using output typehints triggers NotImplementedError('BEAM-2717)' on streaming
```
Comment (Ardagan) [inline tfx_bsl/beam/run_inference.py]: You can also add the commented-out typehint line that should be added (uncommented).

File: tfx_bsl/beam/run_inference.py (hunk 1 above)
Comment (kamilwu) [reply tfx_bsl/beam/run_inference.py]: @Ardagan Sorry, I missed your comment. That's no longer true, the output type in all cases is `PredictionLog` now.

**Comment by kamilwu** [comment]
I've noticed that my commits were merged - thanks! Does it mean I can safely close this pull request?

**Comment by aaltay** [comment]
> I've noticed that my commits were merged - thanks! Does it mean I can safely close this pull request?

I believe the answer is yes. But to be sure, @zhitaoli @rose-rong-liu - could you reply to this?

**Comment by zhitaoli** [comment]
Yes I believe I believe we can close this.

**Comment by kamilwu** [comment]
Closing then, thank you once again for your cooperation
